In [2]:
import pandas as pd
import numpy as np

In [ ]:
##https://github.com/CAIS380-ML-S24/hw02-steven

In [37]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

In [38]:
pip list

Package                       Version
----------------------------- ---------------
aiobotocore                   2.4.2
aiofiles                      22.1.0
aiohttp                       3.8.3
aioitertools                  0.7.1
aiosignal                     1.2.0
aiosqlite                     0.18.0
alabaster                     0.7.12
anaconda-anon-usage           0.4.3
anaconda-catalogs             0.2.0
anaconda-client               1.12.0
anaconda-cloud-auth           0.1.4
anaconda-navigator            2.5.1
anaconda-project              0.11.1
anyio                         3.5.0
appdirs                       1.4.4
argon2-cffi                   21.3.0
argon2-cffi-bindings          21.2.0
arrow                         1.2.3
astroid                       2.14.2
astropy                       5.1
asttokens                     2.0.5
async-timeout                 4.0.2
atomicwrites                  1.4.0
attrs                         22.1.0
autograd                      1.6.2
autograd-

In [24]:
phil = pd.read_csv("phil.csv.csv")
phil.dropna()
#phil.head

Year Punxsutawney Phil  February Average Temperature  \
9    1895         No Record                         26.60   
10   1896         No Record                         35.04   
11   1897         No Record                         33.39   
12   1898       Full Shadow                         35.37   
13   1899         No Record                         25.50   
..    ...               ...                           ...   
126  2012       Full Shadow                         37.51   
127  2013         No Shadow                         34.77   
128  2014       Full Shadow                         32.13   
129  2015       Full Shadow                         32.99   
130  2016         No Shadow                         39.47   

     February Average Temperature (Northeast)  \
9                                        15.6   
10                                       22.2   
11                                       23.6   
12                                       24.8   
13                                       18.1   
..                                        ...   
126                                      30.6   
127                                      25.3   
128                                      21.0   
129                                      12.1   
130                                      28.1   

     February Average Temperature (Midwest)  \
9                                      21.9   
10                                     33.5   
11                                     34.7   
12                                     33.3   
13                                     22.2   
..                                      ...   
126                                    38.6   
127                                    33.4   
128                                    27.1   
129                                    23.4   
130                                    36.9   

     February Average Temperature (Pennsylvania)  March Average Temperature  \
9                                           17.0                      39.97   
10                                          26.6                      38.03   
11                                          27.9                      38.79   
12                                          26.7                      41.05   
13                                          20.0                      37.63   
..                                           ...                        ...   
126                                         34.3                      50.41   
127                                         27.4                      40.91   
128                                         22.8                      40.51   
129                                         16.1                      45.39   
130                                         30.8                      47.50   

     March Average Temperature (Northeast)  \
9                                     27.6   
10                                    25.3   
11                                    32.0   
12                                    38.0   
13                                    29.3   
..                                     ...   
126                                   43.4   
127                                   32.7   
128                                   26.0   
129                                   27.8   
130                                   38.8   

     March Average Temperature (Midwest)  \
9                                   40.2   
10                                  36.9   
11                                  44.0   
12                                  46.0   
13                                  38.4   
..                                   ...   
126                                 56.3   
127                                 37.3   
128                                 38.0   
129                                 41.6   
130                                 49.5   

     March Average Temperature (Pennsylvania)  
9                                      

In [27]:
filtered_data = phil[(phil['Punxsutawney Phil'].isin(['Full Shadow', 'No Shadow'])) & 
                          phil.iloc[:, 2:].notna().all(axis=1)]
print(filtered_data)

     Year Punxsutawney Phil  February Average Temperature  \
12   1898       Full Shadow                         35.37   
14   1900       Full Shadow                         30.76   
15   1901       Full Shadow                         29.86   
17   1903       Full Shadow                         28.42   
18   1904       Full Shadow                         31.59   
..    ...               ...                           ...   
126  2012       Full Shadow                         37.51   
127  2013         No Shadow                         34.77   
128  2014       Full Shadow                         32.13   
129  2015       Full Shadow                         32.99   
130  2016         No Shadow                         39.47   

     February Average Temperature (Northeast)  \
12                                       24.8   
14                                       21.4   
15                                       16.0   
17                                       24.5   
18                    

In [34]:

filtered_data = filtered_data.copy()

filtered_data['Prediction'] = filtered_data['Punxsutawney Phil'].map({'Full Shadow': 0, 'No Shadow': 1})

print(filtered_data['Prediction'])

12     0
14     0
15     0
17     0
18     0
      ..
126    0
127    1
128    0
129    0
130    1
Name: Prediction, Length: 115, dtype: int64


In [35]:
X_overall = filtered_data[['February Average Temperature']]
X_northeast = filtered_data[['February Average Temperature (Northeast)']]
X_midwest = filtered_data[['February Average Temperature (Midwest)']]
X_pennsylvania = filtered_data[['February Average Temperature (Pennsylvania)']]
y = filtered_data['Prediction']

In [55]:
X_midwest.head

<bound method NDFrame.head of      February Average Temperature (Midwest)
12                                     33.3
14                                     27.5
15                                     26.2
17                                     31.3
18                                     28.2
..                                      ...
126                                    38.6
127                                    33.4
128                                    27.1
129                                    23.4
130                                    36.9

[115 rows x 1 columns]>

In [58]:
from sklearn.metrics import precision_score, recall_score, confusion_matrix

In [71]:
def fit_and_evaluate_model(X, y, region_name):
  
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
  
    model = LogisticRegression()
    model.fit(X_train, y_train)
    
   
    y_pred = model.predict(X_test)
   
    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred, zero_division=0)  
    recall = recall_score(y_test, y_pred)
   
    cm = confusion_matrix(y_test, y_pred)
    
   
    print(f"--- Model Summary for {region_name} ---")
    print(f"Accuracy: {accuracy:.4f}")
    print(f"Precision: {precision:.4f}")
    print(f"Recall: {recall:.4f}")
    print("Confusion Matrix:")
    print(cm)
    print(f"Coefficients: {model.coef_[0]}")
    print("Intercept:", model.intercept_[0])
    print("--------------------------------------\n")

In [72]:
fit_and_evaluate_model(X_overall, y, "Overall")

--- Model Summary for Overall ---
Accuracy: 0.8696
Precision: 0.0000
Recall: 0.0000
Confusion Matrix:
[[20  0]
 [ 3  0]]
Coefficients: [0.17663194]
Intercept: -8.04789096742308
--------------------------------------



In [73]:
fit_and_evaluate_model(X_northeast, y, "Northeast")

--- Model Summary for Northeast ---
Accuracy: 0.8696
Precision: 0.0000
Recall: 0.0000
Confusion Matrix:
[[20  0]
 [ 3  0]]
Coefficients: [0.03794997]
Intercept: -2.788171902906656
--------------------------------------



In [74]:
fit_and_evaluate_model(X_midwest, y, "Midwest")

--- Model Summary for Midwest ---
Accuracy: 0.8696
Precision: 0.0000
Recall: 0.0000
Confusion Matrix:
[[20  0]
 [ 3  0]]
Coefficients: [0.04993263]
Intercept: -3.5833607498423476
--------------------------------------



In [75]:
fit_and_evaluate_model(X_pennsylvania, y, "Pennsylvania")

--- Model Summary for Pennsylvania ---
Accuracy: 0.8696
Precision: 0.0000
Recall: 0.0000
Confusion Matrix:
[[20  0]
 [ 3  0]]
Coefficients: [0.03440634]
Intercept: -2.840268742299015
--------------------------------------




Based on the model summaries for various regions and the overall performance, it's evident that Phil has a high accuracy rate of 0.8696 in predicting the arrival of spring. However, the uniform confusion matrices across all regions, showing a failure to correctly predict any positive cases of early spring, highlight a significant limitation in the model's predictive capability. Despite this high level of accuracy, the inability to accurately identify instances of early spring suggests that while Phil's predictions are generally reliable, they lack the specificity needed to effectively forecast early spring occurrences.